### PR TITLE
Machine goes full sparse arrays

### PR DESCRIPTION
**Pull request by jorgemarpa**
Upgrading `machine` to be fully functional with `scipy.sparse` arrays. 
This took some refactoring of how to create `dra`, `ddec`, `r`, and `phi` attributes and how they are used.

The method `_get_source_mask()` has the most changes depending on if using sparse or non-sparse arrays.

The current implementation works ok. All `machine` methods work, but is not optimal:
- Creating sparse arrays can be improved by replacing the for-loop with a tiled version
- Evaluate `<` for sparse arrays is inefficient. I implemented a better option but could be improved when the evaluation is done against an array of values (not a fixed value).

**TODO**

- [x] Unit tests
- [x] evaluate when non-sparse breaks and sparse is inefficient according to N sources and N pixels
- [x] profile functions

### CODE REVIEW

**Comment by jorgemarpa** [comment]
Profiling results are that sparse arrays are efficient when `n_sources * N-pixels > 1e7`.
Test both implementations (numpy and sparse) for Kepler's TPF stacks, cluster mask, and FFIs.


image

File: src/psfmachine/machine.py [hunk 1]
```diff
@@ -749,7 +857,7 @@ def plot_time_model(self):
         cbar.set_label("Normalized Flux")
         return fig
 
-    def build_shape_model(self, plot=False, flux_cut_off=1):
+    def build_shape_model(self, plot=False, flux_cut_off=1, **kwargs):
```
Comment (jorgemarpa) [inline src/psfmachine/machine.py:860]: add `kwargs` so we can pass custom parameters to `_get_source_mask` if necessary.

File: src/psfmachine/machine.py [hunk 2]
```diff
@@ -956,13 +1067,13 @@ def plot_shape_model(self, radius=20):
...
-            self.uncontaminated_source_mask.multiply(self.dra.value),
-            self.uncontaminated_source_mask.multiply(self.ddec.value),
+            self.uncontaminated_source_mask.multiply(self.dra),
+            self.uncontaminated_source_mask.multiply(self.ddec),
         )
         dx = dx.data * u.deg.to(u.arcsecond)
         dy = dy.data * u.deg.to(u.arcsecond)
 
-        fig, ax = plt.subplots(2, 2, figsize=(8, 6.5))
+        fig, ax = plt.subplots(2, 2, figsize=(9.5, 6.5))
```
Comment (jorgemarpa) [inline src/psfmachine/machine.py:1076]: improve figure by avoiding y label overlap

File: src/psfmachine/machine.py [hunk 3]
```diff
@@ -1028,7 +1139,8 @@ def plot_shape_model(self, radius=20):
             xlim=(-radius, radius),
             ylim=(-radius, radius),
         )
-
+        ax[0, 0].set_aspect("equal", adjustable="box")
```
Comment (jorgemarpa) [inline src/psfmachine/machine.py:1142]: keeps ddec vs dra plot with a square aspect ratio

File: src/psfmachine/machine.py [hunk 4]
```diff
@@ -1028,7 +1139,8 @@ def plot_shape_model(self, radius=20):
...
+        ax[1, 0].set_aspect("equal", adjustable="box")
```
Comment (jorgemarpa) [inline src/psfmachine/machine.py:1143]: keeps ddec vs dra plot with a square aspect ratio

File: src/psfmachine/machine.py (hunk 1 above)
Comment (christinahedges) [reply src/psfmachine/machine.py:860]: Is it better to have a specific keyword? That way we can have it in a docstring.

File: src/psfmachine/machine.py [hunk 5]
```diff
@@ -1132,3 +1245,38 @@ def fit_model(self, fit_va=False):
             self.werrs[:, nodata] *= np.nan
 
         return
+
+
+def sparse_lessthan(arr, limit):
+    """
+    Compute less than operation on sparse array by evaluating only non-zero values
+    and reconstructing the sparse array.
+
```
Comment (christinahedges) [inline src/psfmachine/machine.py:1264]: Can you add a quick comment explaining why we need a function for this, and how this function works in a few words?

Comment (christinahedges) [reply src/psfmachine/machine.py:1264]: might be a comment in the other file, you could move that here

**Review by christinahedges — APPROVED** [review]
Hi @jorgemarpa this looks fantastic, I can't find any serious problems here. If you move that comment around I'll get this merged!!

**Comment by jorgemarpa** [comment]
Awesome! I added a few words explaining how that sparse function works and why is needed.

**Comment by christinahedges** [comment]
Thanks for this @jorgemarpa it's great work! I'm merging!

**Pull request merged**